capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -202,9 +450,485 @@ impl<'a> NonvolatileStorage<'a> {
+                // hasn't been allocated one yet
+                if app.has_requested_region && app.region.is_none() {
+                    let region = AppRegion {
+                        // Have this region start where all the existing regions end.
+                        // Note that the app's actual region starts after the region header.
+                        offset: new_header_addr + REGION_HEADER_LEN,
+                        // new regions get handed the same size. this can be
+                        // configured when the capsule is created.
+                        length: self.app_region_size,
+                    };
+
+                    // fail if new region is outside userpace area
+                    if region.offset > self.userspace_start_address + self.userspace_length
+                        || region.offset + region.length
+                            > self.userspace_start_address + self.userspace_length
+                    {
+                        return Err(ErrorCode::FAIL);
+                    }
+
+                    let header = AppRegionHeader {
+                        shortid: shortid.get(),
+                        length: region.length,
+                    };
+
+                    // write this new region header to the end of the existing ones
+                    self.write_region_header(processid, &header, new_header_addr)
+                } else {
+                    // this app never requested to be allocated or its
+                    // region was already allocated
+                    Ok(())
+                }
+            })
+            .unwrap_or(Err(ErrorCode::FAIL))
+    }
+
+    fn check_magic_header(&self) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!("[NONVOLATILE_STORAGE_DRIVER]: Checking magic header");
+        }
+        // kick off a read that reads the first few bytes of the entire userspace
+        // region for the magic header value
+        self.enqueue_command(
+            NonvolatileCommand::HeaderRead(HeaderReadAction::ReadingMagicHeader),
+            self.userspace_start_address,
+            MAGIC_HEADER_LEN,
+            None,
+        )
+    }
+
+    fn write_magic_header(&self) -> Result<(), ErrorCode> {
+        if DEBUG {
+            debug!("[NONVOLATILE_STORAGE_DRIVER]: Writing magic header");
+        }
+
+        let magic_header_slice = u32_to_u8_slice(MAGIC_HEADER);
+        self.header_buffer.map_or(Err(ErrorCode::NOMEM), |buf| {
+            // copy magic value to static buffer
+            for (i, c) in buf[0..magic_header_slice.len()].iter_mut().enumerate() {
+                *c = magic_header_slice[i];
+            }
+            Ok(())
+        })?;
+
+        self.enqueue_command(
+            NonvolatileCommand::HeaderWrite(HeaderWriteAction::WritingMagicHeader),
+            self.userspace_start_address,
+            magic_header_slice.len(),
+            None,
+        )
```

Comment:
Even writing the magic header doesn't erase the flash.

Are you testing this against non-flash storage? We should probably make that clear then

Reply:
> Even writing the magic header doesn't erase the flash.

Should the entire userspace flash area be erased when writing the valid magic header value or is it enough to just erase flash right before we hand it off to apps?

> Are you testing this against non-flash storage? We should probably make that clear then

No, I've only tested with flash storage. Should I clarify this at the top of the file?

Reply:
This isn't flash specific because we are using the NonvolatileStorage HIL. This would work just fine with FRAM, for example.

Reply:
> Should the entire userspace flash area be erased when writing the valid magic header value or is it enough to just erase flash right before we hand it off to apps?

I don't think you should call erase. But it might be worth clearing the region before passing it to an app. You don't know what previous data was stored there and if we should expose it to an application

> No, I've only tested with flash storage. Should I clarify this at the top of the file?

Dose the nRF allow writes without erase? Is the flash already erased? `NonvolatileToPages` seems to never call erase, so this probably won't work on most flash devices.

That's not really something for this PR to fix, but maybe worth documenting somewhere

Reply:
> I don't think you should call erase. But it might be worth clearing the region before passing it to an app. You don't know what previous data was stored there and if we should expose it to an application

Clearing the region makes sense and wouldn't need much code change. I already zero out the region header of the terminating region of the list, so I would just need to zero out the entire region before handing it to an app.

However, what do you think about pushing this off to a different PR that focuses more on freeing regions and reallocating them to different apps (as per the discussion here https://github.com/tock/tock/pull/4109#discussion_r1729477724)?

> Dose the nRF allow writes without erase? Is the flash already erased? NonvolatileToPages seems to never call erase, so this probably won't work on most flash devices.
> That's not really something for this PR to fix, but maybe worth documenting somewhere

I'm not too familiar with the inner workings of the nRF flash chip, but the `nrf52` `nvmc.rs` file seems to do some erasing before writing:

https://github.com/tock/tock/blob/8cef057ca61d941351912ae0ade63319a70bf913/chips/nrf52/src/nvmc.rs#L313-L319

I'm pretty sure this gets used since that same `nvmc` gets passed into the nonvolatile storage driver in the `main.rs` here: 

https://github.com/tock/tock/blob/8cef057ca61d941351912ae0ade63319a70bf913/boards/tutorials/nrf52840dk-thread-tutorial/src/main.rs#L168-L180

Reply:
> However, what do you think about pushing this off to a different PR that focuses more on freeing regions and reallocating them to different apps (as per the discussion here [#4109 (comment)](https://github.com/tock/tock/pull/4109#discussion_r1729477724))?

I think freeing regions is one thing. And can be delayed until later. When a region is freed it will also need to be cleared.

But I also think it's worth erasing a region before we pass it to userspace. We don't know what was previously there and it's possible it's something that we don't wan to expose.

Reply:
I opened https://github.com/tock/tock/issues/4152 about the lack of erase in `NonvolatileToPages`. It's a separate issue to this PR, but note that this implementation won't work on all Tock boards

Reply:
> But I also think it's worth erasing a region before we pass it to userspace. We don't know what was previously there and it's possible it's something that we don't wan to expose.

Region content is now erased after a new region's header is written. https://github.com/tock/tock/pull/4109/commits/80714d383fc3a545b9c83d5c2f1a3d703e3330c1